bpf cleanup

jrajahalme opened the pull request:
Clean up dead code and confusion from the bpf sources. The main change is the reversal of source and destination addresses in the conntrack key. The ports were already reversed, so it was confusing that the addresses and ports were in different order. The fields are also swapped so there is no change in the bpf map, so there should be no upgrade impact.

Please review by commit.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/5848)
<!-- Reviewable:end -->

aanm reviewed (commented):
I find odd that all of these changes don't require a single user-space change in the go code. :thinking:

jrajahalme committed:
bpf: Remove TUPLE_F_OUT

TUPLE_F_OUT was 0, which does not define any flags. It is more honest
to use 0 for "no flags set" when needed.

Remove related stale comments about the single address in the CT
tuple, as the CT tuple now contains both source and destination
addresses.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
bpf: Remove unused cidr parameters

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
bpf: Remove unused ct4_cilium_dbg_tuple()

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
bpf: Skip policy lookup and proxy redirect for reply traffic.

We already ignored the policy lookup results and did not redirect to
proxy if the packet was a reply or related, so it should be safe to
skip policy lookup altogether in this case.

The policy lookup in this case was using the wrong port number due to
the tuple not being reversed when reply direction lookup
succeeds. This did not matter as the policy verdict was ignored for
replies, but it was wrong either way.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
bpf: Swap source and destination addresses in CT key definitions.

CT key addresses are currently set from the corresonding IP header
fields, but the ports are loaded into the key in the reverse
order. Before creating a new CT entry the key is always reversed in
the CT lookup, so that the CT entries now have the addresses in the
reply order, but the ports are in the original direction order.

Fix this by swapping the address fields in the CT entry key and
loading the addresses in to the key from the packet in the reverse
order. This way the CT keys in the bpf maps do not change at all, but
the bpf source code makes more sense, even if it is weird using the
tuple source address as the service destination.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

joestringer commented on bpf/bpf_lxc.c:
```diff
@@ -907,19 +909,21 @@ ipv4_policy(struct __sk_buff *skb, int ifindex, __u32 src_label, int *forwarding
 	csum_l4_offset_and_flags(tuple.nexthdr, &csum_off);
 	is_fragment = ipv4_is_fragment(ip4);
 
+	/* Lookup reverses the tuple if an exiting entry in the reverse direction is not found. */
```
s/exiting/existing

joestringer commented on bpf/lib/common.h:
```diff
@@ -349,16 +357,17 @@ static inline union v6addr *
 }
 
+/* Tuple addresses and ports are loaded from a packet in the reverse order.
+ * For outgoing connections, lib/conntrack swaps them to the correct order. */
```
I wonder if it's worth expanding on this a bit more (+ipv6 equivalent).

Basically we've ended up in a situation where the `ipv4_ct_tuple` in memory never corresponds directly, byte-by-byte, to the contents of a packet we've seen.

When comparing the names of the fields and ordering, the addresses are the same order between `struct ipv4_ct_tuple` and `struct iphdr`, however we load them in the reverse order as follows (from above in this commit):

```
+       tuple.daddr = ip4->saddr;
+       tuple.saddr = ip4->daddr;
```

On the other hand, the ports fields of `struct ipv4_ct_tuple` are in the opposite order of `struct udphdr`, but we load these directly byte-by-byte:

https://github.com/cilium/cilium/blob/71a863422a43e0499cd9b777b68261227544de2a/bpf/lib/conntrack.h#L390-L391

This might be relevant background for any new usages of the tuple, but is unlikely to catch developers out unless they attempt to do byte-wise comparison/copying of the fields.

We could also mention that this is now used across Cilium restarts so the ordering cannot be modified.

Another comment that would be useful is the description of the field names with respect to packet contents, eg:

> The source/destination fields in this structure indicate the source/destination addresses/ports of the original direction of the connection. Users typically populate the ipv4_ct_tuple with the reverse direction first (ie, populate 'saddr' with the packet's 'daddr', 'sport' with 'dport', etc., attempt a CT look up, then if this fails (indicating that there is no corresponding original direction connection), reverse the tuple which makes the sources/destinations match the packet's sources/destinations. At this point, the CT entry will be committed (modulo policy), which means that subsequent replies, following the same logic as above, would find this newly created entry.

joestringer reviewed (changes requested):
Thanks for the clean up, I think it's a step in the right direction.

I put a suggestion below to expand the comments a bit further on the ct tuple structures, which would hopefully elaborate why we're in this situation and how it might affect developers.
